File: day8/day8.ts
const fs = require('fs');
import { count, Dictionary, distinct, filter, filtermap, flatmap, groupbytoobject, pipe, toarray } from "powerseq";

type Point = { x: number; y: number };
type FrequencyMap = Dictionary<Point[]>;

function isInsideMap(width: number, height: number, point: Point) {
	return point.x >= 0 && point.x < width && point.y >= 0 && point.y < height;
}

function parseInput(input: string): FrequencyMap {
	return pipe(
		input.split('\n'),
		flatmap((r, x) =>
			pipe(
				r.split(''),
				filtermap((el, y) => el !== '.' ? { x, y, value: el } : null),
			)),
		groupbytoobject(el => el.value, el => ({ x: el.x, y: el.y }))

	)
}

function calculateAntinodes(a: Point, b: Point): Point[] {
	const distanceX = b.x - a.x;
	const distanceY = b.y - a.y;
	return [
		{ x: a.x - distanceX, y: a.y - distanceY },
		{ x: b.x + distanceX, y: b.y + distanceY }
	];
}

function calculateAllAntinodes(antennas: Point[], extended?: boolean, width?: number, height?: number): Point[] {
	return pipe(
		antennas,
		flatmap((p1, i) => pipe(
			antennas,
			filter((_, j) => j > i),
			flatmap(p2 => !extended ? calculateAntinodes(p1, p2) : calculateExtendedAntinodes(p1, p2, width, height))
		)),
		toarray()
	);
}

function calculateAntinodesCount(input: string, extended?: boolean) {
	const antennas = parseInput(input);
	const rows = input.split('\n');

	return pipe(
		Object.values(antennas),
		flatmap(entry => !extended
			? calculateAllAntinodes(entry)
			: calculateAllAntinodes(entry, true, rows[0].length, rows.length)
		),
		filter(p => isInsideMap(rows[0].length, rows.length, p)),
		distinct(p => `${p.x},${p.y}`),
		count()
	)
}

//1.
function part1(input: string) {
	return calculateAntinodesCount(input);
}

//2.
function calculateSingleAntinode(a: Point, b: Point, distanceX: number, distanceY: number, dir: 'A' | 'B'): Point {
	return dir === 'A' ? { x: a.x - distanceX, y: a.y - distanceY } : { x: b.x + distanceX, y: b.y + distanceY };
}

function calculateExtendedAntinodes(a: Point, b: Point, width: number, height: number) {
	function calculateExtendedAntinodesInnerIterative(start: Point, next: Point, dir: 'A' | 'B'): Point[] {
		const result: Point[] = [a, b];
		let current = next;
		let isOutsideMap = false;
		const distanceX = b.x - a.x;
		const distanceY = b.y - a.y;
		while (!isOutsideMap) {
			const antinode = dir === 'A'
				? calculateSingleAntinode(start, current, distanceX, distanceY, 'A')
				: calculateSingleAntinode(start, current, distanceX, distanceY, 'B');

			if (!isInsideMap(width, height, antinode)) {
				isOutsideMap = true;
			} else {
				result.push(antinode);
				start = current;
				current = antinode;
			}
		}
		return result;
	}

	return [
		...calculateExtendedAntinodesInnerIterative(a, b, 'A'),
		...calculateExtendedAntinodesInnerIterative(a, b, 'B')
	];
}

function part2(input: string) {
	return calculateAntinodesCount(input, true);
}

const rawInput = fs.readFileSync('./day8/input.txt', 'utf-8');
console.log(part1(rawInput));
console.log(part2(rawInput));

File: day8/day8.go
package day8

import (
	"aoc2024/aochelpers"
	"fmt"
	"strings"
)

type Point struct {
	X, Y int
}

type FrequencyMap map[rune][]Point

func isInsideMap(width, height int, point Point) bool {
	return point.X >= 0 && point.X < width && point.Y >= 0 && point.Y < height
}

func parseInput(input string) FrequencyMap {
	freqMap := make(FrequencyMap)
	lines := strings.Split(input, "\n")

	for x, line := range lines {
		for y, ch := range line {
			if ch != '.' {
				freqMap[ch] = append(freqMap[ch], Point{x, y})
			}
		}
	}
	return freqMap
}

func calculateAntinodes(a, b Point) []Point {
	distanceX := b.X - a.X
	distanceY := b.Y - a.Y
	return []Point{
		{X: a.X - distanceX, Y: a.Y - distanceY},
		{X: b.X + distanceX, Y: b.Y + distanceY},
	}
}

func calculateAllAntinodes(antennas []Point, extended bool, width, height int) []Point {
	var results []Point

	for i, p1 := range antennas {
		for j := i + 1; j < len(antennas); j++ {
			p2 := antennas[j]
			var newAntinodes []Point
			if(extended){
				newAntinodes = calculateExtendedAntinodes(p1, p2, width, height)
			} else {
				newAntinodes = calculateAntinodes(p1, p2)
			}
			results = append(results, newAntinodes...)
		}
	}
	return results
}

func calculateAntinodesCount(input string, extended bool) int {
	antennas := parseInput(input)
	rows := strings.Split(input, "\n")
	width, height := len(rows[0]), len(rows)

	pointSet := make(map[string]bool)

	for _, points := range antennas {
		antinodes := calculateAllAntinodes(points, extended, width, height) 
		for _, p := range antinodes {
			if isInsideMap(width, height, p) {
				key := fmt.Sprintf("%d,%d", p.X, p.Y)
				pointSet[key] = true
			}
		}
	}
	return len(pointSet)
}

func Part1() {
	input := aochelpers.ReadFileAsString("./day8/input.txt")
	result := calculateAntinodesCount(input, false)
	fmt.Println("Day 8, part 1:", result)
}

//2.
func calculateSingleAntinode(a, b Point, distanceX, distanceY int, dir rune) Point {
	if( dir == 'A' ){
		return Point{ X: a.X - distanceX, Y: a.Y - distanceY } 
	 } else {
		return Point{ X: b.X + distanceX, Y: b.Y + distanceY }
	 }
}

func calculateExtendedAntinodes(a,b Point, width, height int) []Point {
	var calculateExtendedAntinodesInnerIterative func(start, next Point, dir rune) []Point
	
	calculateExtendedAntinodesInnerIterative = func(start, next Point, dir rune) []Point {
		result := []Point{a,b}
		current := next
		isOutsideMap := false
		distanceX := b.X - a.X
		distanceY := b.Y - a.Y
		for !isOutsideMap {
			var antinode Point;
			if(dir == 'A'){
				antinode = calculateSingleAntinode(start, current, distanceX, distanceY, 'A')
			} else {
				antinode = calculateSingleAntinode(start, current, distanceX, distanceY, 'B')
			}
			if(!isInsideMap(width, height, antinode)){
				isOutsideMap = true
			} else {
				result = append(result, antinode)
				start = current
				current = antinode
			}
		}
		return result
	}

	part1 := calculateExtendedAntinodesInnerIterative(a, b, 'A')
	part2 := calculateExtendedAntinodesInnerIterative(a, b, 'B')
	return append(part1, part2...)
}

func Part2() {
	input := aochelpers.ReadFileAsString("./day8/input.txt")
	result := calculateAntinodesCount(input, true)
	fmt.Println("Day 8, part 2:", result)
}